Support simulated time

skoudmar opened the pull request:
Simulated time can be enabled:

- Implicitly, by registering a parameter handler. While launching the node, parameter `use_sim_time` must be set to `true`.
- Explicitly, by calling `TimeSource::enable_sim_time(&self)`.



It works but needs more polishing.

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -7,6 +7,7 @@ use std::collections::HashMap;
 use std::ffi::{CStr, CString};
 use std::future::Future;
 use std::mem::MaybeUninit;
+use std::pin::Pin;
```
This related to the commit message:
> During the initialization of timer with RosTime clock the timer adds jump callback with rcl_jump_callback_info_t to the clock with .user_data being the address of the timer handle.

This is not clear? Which initialization do you mean? In r2r or in rcl? Where is the "jump callback" (btw what it is?) added? Name the exact function or method you are talking about. "To the clock with ..." - again - be specific. Do you mean some specific data structure? Name it exactly.

Also write in which function would the segfault happen.

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -1122,17 +1127,62 @@ impl Node {
+        let out_timer = Timer { receiver: rx };
+
+        Ok(out_timer)
+    }
```
This is confusing. It seems you are adding this code, but you are actually changing some part of it. The previous version is two pages below so it took me non-trivial amount of time to figure out what was actually changes and why.

It seems that this is just due to making the `_clock` Optional. Please make this change to `Optional<Clock>` a separate commit. This will make reading this commit much easier.

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -1172,7 +1211,7 @@ impl Node {
 struct Timer_ {
     timer_handle: Pin<Box<rcl_timer_t>>,
-    _clock: Clock, // just here to be dropped properly later.
+    _clock: Option<Clock>, // Some(clock) if the timer owns the clock, just here to be dropped properly later.
```
As noted above, make this separate commit (before this one).

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -53,6 +53,8 @@ pub struct Node {
     timers: Vec<Timer_>,
     // and the publishers, whom we allow to be shared.. hmm.
     pubs: Vec<Arc<rcl_publisher_t>>,
+    // shared ros clock
```
Shared between what? Perhaps change to "RosTime clock used by all timers created by create_timer()

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -53,6 +53,8 @@ pub struct Node {
     timers: Vec<Timer_>,
```
I thing the commit message can start like this:

> Allow creating timers of type ClockType::RosTime
> 
> This adds Node::create_timer() method. The method is similar to
> already existing Node::create_wall_timer(), but the timer is based on
> ClockType::RosTime rahter than on ClockType::SteadyTime.
> 
> The advantage of ClockType::RosTime is that one can drive the clock
> from a different time source. This will be used in the next commits to
> implement for clock source on the /clock topic.
>

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -52,6 +55,10 @@ pub struct Node {
     pubs: Vec<Arc<rcl_publisher_t>>,
+    // shared ros clock
+    pub(crate) ros_clock: Arc<Mutex<Clock>>,
+    // time source that provides simulated time
```
```suggestion
    // time source that provides time from the /clock topic
```

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -1207,6 +1237,19 @@ impl Node {
+
+    pub fn get_ros_clock(&self) -> Arc<Mutex<Clock>> {
+        self.ros_clock.clone()
+    }
```
These public functions should be documented.

wentasah commented on r2r/src/nodes.rs:
```diff
@@ -1207,6 +1237,19 @@ impl Node {
+        self.ros_clock.clone()
+    }
+
+    pub fn now(&self) -> Result<Duration> {
```
This function doesn't seem to be used anywhere. Use it at least in some example (and document it) or remove it.

wentasah commented on r2r/Cargo.toml:
```diff
@@ -47,6 +47,7 @@ prettyplease = "0.2.6"
 [features]
```
Commit message:

You should first say what you are doing here and only then mention that it's feature-gated and why.

As far as I can see you are:
1. Adding TimeSource type. Mention that is related (or implements the same thing) as some rclcpp type.
2. Adding `Clock` methods `*_time_override`. Again, mention what are their rclcpp equivalents.
3. Initializing TimeSource in every node.
4. Enabling the simulated time if Node parameter `use_sim_time` is true.
5. Adding some `Node` methods.

It is a question, whether we need all these changes in a single commit. If this would be split to smaller pieces, it would be easier to figure out which change is relate to which of the above 5 points.

wentasah commented on r2r/examples/sim_time_publisher.rs:
```diff
@@ -0,0 +1,30 @@
+        msg.clock = Clock::to_builtin_time(&time);
+
+        publisher.publish(&msg)?;
+        println!("Publishing time {:3} s {:9} ns", time.as_secs(), time.subsec_nanos());
```
```suggestion
        println!("Publishing time {}.{:9} s", time.as_secs(), time.subsec_nanos());
```

wentasah commented on r2r/src/clocks.rs:
```diff
@@ -65,12 +69,61 @@ impl Clock {
+        Ok(())
+    }
+
+    pub(crate) fn disable_ros_time_override(&mut self) -> Result<()> {
```
When compiling without `sim-time` feature, I get a lot of  **never used** warnings, such as:

    warning: methods `disable_ros_time_override` and `set_ros_time_override` are never used

wentasah commented on r2r/examples/timer_sim_time.rs:
```diff
@@ -0,0 +1,69 @@
+
+        println!("Timer called ({}), {}us since last call", x, elapsed.as_micros());
+        println!(
+            "\tcurrent time ros={}s, system={}s",
```
```suggestion
            "\tcurrent time ros={:.3}s, system={:.3}s",
```

wentasah commented on r2r/examples/timer_sim_time.rs:
```diff
@@ -0,0 +1,69 @@
+    let mut node = r2r::Node::create(ctx, "testnode", "")?;
+
+    let time_source = node.get_time_source();
+    time_source.enable_sim_time(&mut node)?;
```
When I try commenting this line and running the example as:

    cargo run --features=sim-time --example=timer_sim_time -- --ros-args -p use_sim_time:=true
    
sim time is not enabled.

I would prefer not having this line here and having a comment in the source code about how to run it with sim time enabled. Perhaps, if the sim time is not enabled, the example could print how to enable it to stdout.

wentasah commented on r2r/examples/timer_sim_time.rs:
```diff
@@ -0,0 +1,69 @@
+async fn timer_task(
+    mut t: r2r::Timer, ros_clock: Arc<Mutex<Clock>>, mut system_clock: Clock,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let mut x: i32 = 0;
```
Rename `x` to something more descriptive. For example `iteration`.

wentasah commented on r2r/examples/timer_sim_time.rs:
```diff
@@ -0,0 +1,69 @@
+use futures::executor::LocalPool;
```
Add a comment how to run this example. See below and also mention that the functionality can be demonstrated together with `sim_time_publisher` (or with `ros2 bag play --clock` )

wentasah commented on r2r/src/time_source.rs:
```diff
@@ -38,6 +38,15 @@ impl TimeSource {
 
+    /// Attach clock of type [RosTime](ClockType::RosTime) to the [TimeSource]
+    ///
+    /// If the simulated time is enabled the [TimeSource] will distribute simulated time
```
... enabled by calling `enable_sim_time()` ...

wentasah commented on r2r/src/time_source.rs:
```diff
@@ -38,6 +38,15 @@ impl TimeSource {
         }
     }
 
```
Document also the `TimeSource` struct - at least one sentence. Currently the documentation looks like this:

![image](https://github.com/wentasah/r2r/assets/140542/d00dfc9d-510b-4c3f-a208-f70418ee452b)

wentasah commented on r2r/src/time_source.rs:
```diff
@@ -98,6 +112,12 @@ impl TimeSource {
+    ///
+    /// # Arguments
+    ///
+    /// * `remove_subscriber`: Whether to schedule removal of subscriber to the `"/clock"`
+    ///  topic in the next call to [Node::spin_once].
```
You document that happens if `remove_subscriber` is true. But what happens if it's false? Does it have some consequences? Why it is useful to have this parameter at all?

wentasah commented on r2r/src/time_source.rs:
```diff
@@ -38,6 +38,15 @@ impl TimeSource {
         }
     }
 
+    /// Attach clock of type [RosTime](ClockType::RosTime) to the [TimeSource]
```
In general, the documentation should be the part of the commit, which adds the code. It would make the review much easier.

wentasah reviewed (changes requested):
I went through it. It works, but I have a few comments and change requests. Most of them are related to making it easier to understand the code (and commit messages).

BTW I reviewed it commit by commit. Since github doesn't allow commenting on commit messages, I added comments about them to the first lines of the diff. If you look at the comments below or in *Files changed* tab, it would be confusing, because here, it's not visible to which commit the comment belongs.

skoudmar committed:
Changed Timer_ to not require clock ownership

This allows to implement timers that can share single clock among them.

This commit is preparation for feature sim time.

skoudmar committed:
Allow creating timers of type ClockType::RosTime

This adds Node::create_timer() method. The method is similar to
already existing Node::create_wall_timer(), but the timer is based on
ClockType::RosTime rather than on ClockType::SteadyTime.

The advantage of ClockType::RosTime is that one can drive the clock
from a different time source. This will be used in the next commits to
implement for clock source on the /clock topic.

skoudmar committed:
Add examples showing usage of sim time

skoudmar committed:
Enable simulated time when the node is started with use_sim_time parameter

add `--ros-args -p use_sim_time:=true` to starting command line

skoudmar committed:
Add methods to access TimeSource and shared ROS clock

wentasah commented:
Narazili jsme na problém, že když `/clock` poskočí o víc než je 2x perioda timeru, tak se timer callback vyvolá jen jednou a ne víckrát, jak bych očekával. Můžeš zjistit, jestli je to chyba v této implementaci, nebo jestli se `rclcpp` chová stejně?

skoudmar commented:
Takhle je to implementováno už v `rcl` a `rclcpp` chování nijak nemění.

Při každé změně času se timer callback zavolá nejvýše jednou.

tfoldi committed:
chore: use BEST_AVAILABLE QoS in rolling

m-dahl committed:
wait a bit for publisher info in test.

m-dahl committed:
Release 0.8.4

skoudmar committed:
Pin timer_handle to avoid SIGSEGV in _rcl_timer_time_jump() callback

This commit is preparation for next commit with timer using RosTime clock.

In the previous implementation the function rcl_timer_init is called
with zero initialized timer handle being stored on stack and after
initialization with rcl_timer_init the timer handle is moved by rust
inside Timer_ struct.

The problem occurs if the type of clock being passed to rcl_timer_init()
is RosTime because the function also registers callbacks in the clock
that would notify the timer about sudden time change (jump callback).
In RCL the jump callback is registered with rcl_clock_add_jump_callback()
and the argument callback.user_data is pointer to timer handle that in
our case references timer stored om stack. When returning from function
create_wall_timer() the pointer becomes dangling and therefore when the
callback is actually called (by function _rcl_timer_time_jump) it could
result in SIGSEGV or could access some variable that happens to be in
the same place as the timer_handle was on stack.

This is fixed by storing timer_handle inside pinned box to prevent
moving the handle.

skoudmar committed:
Initialize TimeSource in node

skoudmar committed:
Add time_source module to manage RosTime clocks

The TimeSource subscribes to /clock topic and publishes the time to all
attached clocks (by default only clock used by create_timer is attached).

This module is feature gated because the clock is published in message
rosgraph_msgs::msg::Clock and rosgraph_msgs needs to be specified as
dependency by the end users.

skoudmar committed:
Prepare clock to support simulated time

Added feature flag sim-time because functions that are feature gated does
not make sense without TimeSource that is implemented in later commit
and needs the feature flag.

m-dahl committed:
Replace feature with cfg based on whether clock message is compiled

skoudmar committed:
Add panic to sim time examples

skoudmar committed:
Add macro to fail compilation if sim time is not supported

cfg(r2r__rosgraph_msgs__msg__Clock) attribute is not propagated to user
code so this macro can be used to intentionally fail compilation if
sim time is not supported.

m-dahl committed:
Update README.

passchaos committed:
feat: remove is_available's node mut-ref dependency, also remove the !Sync constraint
on the future returned

jerry73204 committed:
Fix several clippy warnings

jerry73204 committed:
Add "edition" and other options in rustfmt.toml

jerry73204 committed:
Run rustfmt using new configuration

jerry73204 committed:
[r2r_common] Remove println!() that taints the build script output

jerry73204 committed:
Add an r2r_info example to dump the ROS environment report

wentasah committed:
Make the order of parameters deterministic

GUI tools for working with parameters (at least Foxglove Studio, rqt
and rig_reconfigure) show the parameters in the same order as returned
by the ListParameters service. r2r stores the parameters in a HashMap,
which iterates keys and values in arbitrary order. The result is that
the GUI tools show the parameters in different order after every node
invocation, which can be quite annoying.

To make the order deterministic, we change the parameter storage from
HashMap to IndexMap, which iterates the map in insertion order.
According to the indexmap documentation, IndexMap is a drop-in
replacement of HashMap so this change should not require code changes
in applications using r2r. At least r2r examples and my projects
needed no changes.

wentasah committed:
Make order of derived parameters independent of CLI arguments

The previous commit made the order of parameters deterministic
(defined by so called insertion order), but the order still depends on
whether or not some parameters were set from the command line. Command
line parameters are inserted to the IndexMap at Node creation time,
while other RosParams-derived parameters later, during the call to
make_derived_parameter_handler().

This commit ensures that parameters processed by
make_derived_parameter_handler() are always inserted in the same
order, even if some of them were already inserted before due to CLI
arguments.

m-dahl committed:
Update README

TijlJappens committed:
Expose qos profiles for service clients and servers. (#95)

Adds a QosProfile parameter for `Node::create_client` and `Node::node.create_service`.

`QosProfile::default()` can be passed when updating existing code.

m-dahl committed:
Temporary fix for CI until #96 is resolved.

m-dahl committed:
Check for null in `from_raw_parts` which rust 1.78 does not like.

Revert temporary fix for CI. Closes #96.

m-dahl committed:
rustfmt

m-dahl committed:
Update README

m-dahl committed:
Release 0.9.0

tobiasstarkwayve committed:
Avoid memory leak in loaned-message return path

des256 committed:
Add ros2 jazzy support.

m-dahl committed:
Remove dead code.

tobiasstarkwayve committed:
Copy topic name before deleting subscription.

This fixes undefined behaviour, as the
rcl_subscription_get_topic_name documentation
states that the topic is no longer valid when
the subscription dies

tobiasstarkwayve committed:
Add additional assertions to rcl wrapper

tobiasstarkwayve committed:
Use copy_nonoverlapping for more efficient sequence filling

m-dahl committed:
Release 0.9.1

IamTheCarl committed:
More convinent one time parameter access (#107)

* Implement TryInto for ParameterValue

* Add `get_parameter` method.

* Consolidate 'ParameterNotSet' into 'ParamterWrongType' error

* Add support for optional parameters

* Add example for get_parameter.

m-dahl committed:
Silence cfg and ffi warnings. Fixes #108.

m-dahl committed:
Update README and fix build docs mistake.

m-dahl committed:
Release 0.9.2

m-dahl committed:
add destroy publisher functions

m-dahl committed:
Update readme

showier-drastic committed:
Expose WrongParameterType

m-dahl committed:
Release 0.9.3

m-dahl committed:
single-':' cargo syntax for compability with rust versions < 1.77

m-dahl committed:
fix message generation for WChar idl type

m-dahl committed:
Release 0.9.4
